library/std/src/io/util.rs

```diff
@@ -268,3 +268,154 @@ impl fmt::Debug for Sink {
         f.debug_struct("Sink").finish_non_exhaustive()
     }
 }
+
+/// A writer which will move data into the void, and a reader which
+/// is always at EOF. Effectively a combination of [`Sink`] and [`Empty`],
+/// similiar to `/dev/null` on Unix.
+///
+/// This struct is generally created by calling [`null()`]. Please
+/// see the documentation of [`null()`] for more details.
+#[unstable(feature = "io_null", issue = "none")]
+#[non_exhaustive]
+#[derive(Copy, Clone, Default)]
+pub struct Null;
+
+/// Creates an instance of a writer which will successfully consume all data, as
+/// well as a reader which is always at EOF.
+///
+/// All calls to [`write`] on the returned instance will return `Ok(buf.len())`
+/// and the contents of the buffer will not be inspected.
+///
+/// All calls to [`read`] from the returned reader will return `Ok(0)`.
+///
+/// [`write`]: Write::write
+/// [`read`]: Read::read
+///
+/// # Examples
+///
+/// ```rust
+/// # #![feature(io_null)]
+/// use std::io::{self, Write};
+///
+/// let buffer = vec![1, 2, 3, 5, 8];
+/// let num_bytes = io::null().write(&buffer).unwrap();
+/// assert_eq!(num_bytes, 5);
+/// ```
+///
+///
+/// ```rust
+/// # #![feature(io_null)]
+/// use std::io::{self, Read};
+///
+/// let mut buffer = String::new();
+/// io::null().read_to_string(&mut buffer).unwrap();
+/// assert!(buffer.is_empty());
+/// ```
+#[must_use]
+#[unstable(feature = "io_null", issue = "none")]
+#[rustc_const_unstable(feature = "const_io_structs", issue = "78812")]
+pub const fn null() -> Null {
+    Null
+}
+
+#[unstable(feature = "io_null", issue = "none")]
+impl Write for Null {
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        Ok(buf.len())
+    }
+
+    #[inline]
+    fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
+        let total_len = bufs.iter().map(|b| b.len()).sum();
+        Ok(total_len)
+    }
+
+    #[inline]
+    fn is_write_vectored(&self) -> bool {
+        true
+    }
+
+    #[inline]
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+#[unstable(feature = "io_null", issue = "none")]
+impl Write for &Null {
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        Ok(buf.len())
+    }
+
+    #[inline]
+    fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
+        let total_len = bufs.iter().map(|b| b.len()).sum();
+        Ok(total_len)
+    }
+
+    #[inline]
+    fn is_write_vectored(&self) -> bool {
+        true
+    }
+
+    #[inline]
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
```

Comment:
when would this be useful?

Reply:
No clue, but it was `impl`'d for `io::Sink`, so I didn't want there to be any weird cases where a user couldn't 1:1 switch from `Sink` to `Null`.